### PR TITLE
Fixed segfaults in GDAL_CG_Create and GDALGridContextFree

### PR DESCRIPTION
**Pull request by albertov**
GDAL_CG_Create was not initializing GDALContourGenerator properly so the first call to GDAL_CG_FeedLine would crash as the last scanline would be NULL.

Also fixed GDALGridContextFree in the case that GDAL_NUM_THREADS is < 1.

### CODE REVIEW

File: gdal/alg/gdalgrid.cpp [hunk 1]
```diff
@@ -2155,7 +2155,8 @@ void GDALGridContextFree(GDALGridContext* psContext)
         CPLFree(psContext->pabyZ);
         if( psContext->sExtraParameters.psTriangulation )
             GDALTriangulationFree( psContext->sExtraParameters.psTriangulation );
-        delete psContext->poWorkerThreadPool;
+        if (psContext->poWorkerThreadPool)
```
Comment (rouault) [inline gdal/alg/gdalgrid.cpp]: Why is this necessary ? delete NULL is OK

Comment (albertov) [inline gdal/alg/gdalgrid.cpp]: I'm not sure really. I've just learnt that `delete NULL` is ok. I was chasing the segfault when running the integration tests of these [haskell bindings](https://github.com/albertov/bindings-gdal) which are multi-threaded so perhaps the segfault was caused by the un-initialized contour generator in another thread. I'll run them again without this change and report back

Comment (albertov) [inline gdal/alg/gdalgrid.cpp]: It is indeed unnecessary. I've dropped the commit that changes this

**Comment by rouault** [comment]
Fixed per https://trac.osgeo.org/gdal/ticket/6491
